cni/pkg/ambient/informers.go

```diff
@@ -120,31 +121,49 @@ func (s *Server) Reconcile(input any) error {
 			s.AddPodToMesh(pod)
 		}
 	case controllers.EventUpdate:
-		// For update, we just need to handle opt outs
-		newPod := event.New.(*corev1.Pod)
-		oldPod := event.Old.(*corev1.Pod)
-		ns := s.namespaces.Get(newPod.Namespace, "")
-		if ns == nil {
-			return fmt.Errorf("failed to find namespace %v", ns)
-		}
-		wasEnabled := oldPod.Annotations[constants.AmbientRedirection] == constants.AmbientRedirectionEnabled
-		nowEnabled := PodRedirectionEnabled(ns, newPod)
-		if wasEnabled && !nowEnabled {
-			log.Debugf("Pod no longer matches, removing from mesh")
-			s.DelPodFromMesh(newPod, event)
-		} else if !wasEnabled && nowEnabled {
-			log.Debugf("Pod now matches, adding to mesh")
-			s.AddPodToMesh(pod)
-		} else if nowEnabled && !IsPodInIpset(pod) {
-			// This can happen if a node cleanup happens as part of a ztunnel recycle,
-			// cleaning up the node-level ipset with all the pod IPs
-			// If this happens we re-queue everything for reconciliation, and need to
-			// make sure existing pods get re-added to the ipset if they aren't already there.
-			log.Debugf("Pod is enabled but not in ipset, (re)adding to mesh")
-			s.AddPodToMesh(pod)
-		}
+		return s.handleUpdate(event)
 	case controllers.EventDelete:
 		s.DelPodFromMesh(pod, event)
 	}
 	return nil
 }
+
+func (s *Server) handleUpdate(event controllers.Event) error {
+	// For update, we just need to handle opt outs
+	newPod := event.New.(*corev1.Pod)
+	oldPod := event.Old.(*corev1.Pod)
+	log := log.WithLabels("type", event.Event, "pod", config.NamespacedName(newPod))
+
+	ns := s.namespaces.Get(newPod.Namespace, "")
+	if ns == nil {
+		return fmt.Errorf("failed to find namespace %v", ns)
+	}
+
+	wasEnabled := oldPod.Annotations[constants.AmbientRedirection] == constants.AmbientRedirectionEnabled
+	nowEnabled := PodRedirectionEnabled(ns, newPod)
+
+	noBusiness := !wasEnabled && !nowEnabled
+	podNeedRemoveToMesh := wasEnabled && !nowEnabled
+	podIsJoiningMesh := !wasEnabled && nowEnabled
+	meshedPodNotInIpset := nowEnabled && !IsPodInIpset(newPod)
+
+	switch {
+	case noBusiness: // no business with ambient mesh
+		return nil
+	case podNeedRemoveToMesh:
+		log.Debugf("Pod no longer matches, removing from mesh")
+		s.DelPodFromMesh(newPod, event)
+	case podIsJoiningMesh:
+		log.Debugf("Pod now matches, adding to mesh")
+		s.AddPodToMesh(newPod)
+	case meshedPodNotInIpset:
+		// This can happen if a node cleanup happens as part of a ztunnel recycle,
+		// cleaning up the node-level ipset with all the pod IPs
+		// If this happens we re-queue everything for reconciliation, and need to
+		// make sure existing pods get re-added to the ipset if they aren't already there.
+		log.Debugf("Pod is enabled but not in ipset, (re)adding to mesh")
+		s.AddPodToMesh(newPod)
```

Comment:
nit: can merge these two cases

Reply:
I prefer to separate them, they has different comments and logs which make the logic clear for developer